### ISSUE TITLE
DPoP/Proof of possession question

### ISSUE THREAD

**Issue by CarterDawson72** [panva/node-openid-client#206]
Simple question not an issue per si:

I'm using the oidc-provider with the openid-client. 

I've seen that oidc-provider allows to have DPoP/Proof of possession enabled, and so I did on the oidc-provider features for the configuration ( dPoP: { enabled: true, iatTolerance: 60, ack: 'id-03' } ) and I can see that now there is a DPoP end point available.

But with the feature enabled on the provider, I can't find any way for the openid-client to request a DPoP based token.

I've tried several things such as creating the required payload for client side generated keys, but no success. The authorization request always returns a Bearer token instead of Beared+DPop token.
Any ideas or tips that can help?

**Comment by panva** [comment]
> I can see that now there is a DPoP end point available.

DPoP Endpoint? There isn't one. What are you seeing?

> But with the feature enabled on the provider, I can't find any way for the openid-client to request a DPoP based token.

They are really different projects, not everything supported on the OP is going to end up being prioritized or implemented for the node client.

The client doesn't allow for headers to be passed to the `callback()` method to be sent to the token_endpoint, which is what would trigger dpop.

For the client I only implement final or stable rfcs, so you're gonna have to wait for a DPoP a bit to be officially supported in the API.

You can however plug into the custom http options API already today.

```js
const { custom } = require('openid-client');
const jose = require('jose');
const nanoid = require('nanoid');

const privateKey = jose.JWK.generateSync('EC');
const publicJWK = privateKey.toJWK(false);
delete publicJWK.kid;

client[custom.http_options] = function (options) {
  if (options.url === client.issuer.token_endpoint && options.body.grant_type === 'authorization_code') {
    const DPoP = jose.JWS.sign({
      htu: options.url,
      htm: options.method,
      iat: Math.floor(Date.now() / 1000),
      jti: nanoid(),
    }, 
    privateKey,
    {
      typ: 'dpop+jwt',
      alg: 'ES256',
      jwk: publicJWK,
    });
    options.headers = options.headers || {};
    options.headers.DPoP = DPoP;
  }

  return options;
}
```

**Comment by CarterDawson72** [comment]
Thanks for the quick answer.

Regarding the DPoP end point, my bad, it was an issue with the express router.

Regarding the request for a DPoP token, thanks for clarifying the issue, and will use your approach.

**Comment by panva** [comment]
DPoP is now supported by the openid-client API, released with v4.1.0
